### ISSUE TITLE
libsmpack ffi/struct_layout.rb:88:in `get': invalid memory read at address=0x00000000000000 (FFI::NullPointerError)

### ISSUE THREAD

**Issue by CAMOBAP** [metanorma/packed-mn#91]
Happens on both linux and macos:
 - https://github.com/metanorma/packed-mn/runs/2344678405
 - https://github.com/metanorma/packed-mn/runs/2345205604

cc @ronaldtse @alexeymorozov

**Comment by alexeymorozov** [comment]
I've managed to fix it locally by switching from Ruby 2.4.1 to 2.4.10. How can it be done in [ruby-packer](https://github.com/metanorma/ruby-packer)?

The bug can be reproduced locally:

```ruby
mkdir tmp
cp Gemfile* tmp
cp bin/metanorma tmp
cp -R vendor tmp
cd tmp
bundle install
git clone --recurse-submodules https://github.com/metanorma/mn-samples-iso iso
./metanorma site generate iso -c iso/metanorma.yml -o site/iso --agree-to-terms
```

**Comment by CAMOBAP** [comment]
@alexeymorozov thanks for sharing this, let me have a look, previously I upgraded bundler to a new one, maybe I can do the same with ruby

**Comment by CAMOBAP** [comment]
Well, ruby upgrade looks much more complex task than bundler upgrade. @alexeymorozov maybe you have some insights what exactly changes in ruby make this working on ruby 2.4.10?

P.S. I'm continue work on ruby upgrade

**Comment by alexeymorozov** [comment]
I've tried to look through commits of Ruby update, but could not find the needed commit/fix. Seems that it happens on the second file in an archive, so some kind of pointer overwrite happens.

I saw there is some work on Ruby 2.6, may be we can use it?

**Comment by alexeymorozov** [comment]
It works on Ruby 2.4.2, [build](https://github.com/alexeymorozov/packed-mn/runs/2399620614).

So we have at least 2 options: compile with Ruby 2.4.2+ or find the needed fix between 2.4.1 and 2.4.2.

**Comment by CAMOBAP** [comment]
@ronaldtse @alexeymorozov 
Update:
 - I will continue work on ruby 2.4.10 migration, I hope it will be done today

**Comment by alexeymorozov** [comment]
Managed to compile [ruby-packer with Ruby 2.4.2 on Linux](https://github.com/alexeymorozov/ruby-packer/raw/packed-mn-2.4.2-patched/rubyc-linux-x64).

Also added [a patch (27 MB)](https://github.com/alexeymorozov/ruby-packer/raw/packed-mn-2.4.2-patched/packed-mn-binary.patch
) between original Ruby and Ruby of metanorma's ruby-packer.

Here is [a PR with Ruby source](https://github.com/alexeymorozov/ruby-packer/pull/1).

For compilation I used a `m5zn.xlarge` instance at AWS. It has Intel Xeon Scalable 4.5 GHz, 16 GB RAM, and eats $0.34/hour. It compiles ruby-packer for Linux for 5:30 minutes, and packed-mn in 14 minutes.

With a new ruby-packer (with 2.4.2 ruby) there is no NullPointerError error.
